protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+all features within that hardfork (e.g., EIPs, protocol changes) activate
+simultaneously. Bundling features in this way simplifies coordination but
+limits flexibility, such as enabling individual features for testing or
+delaying specific ones without forking the config.
+
+In the past, contributors of the OP Stack have had issues surrounding
+hardfork features. For example, long before the Isthmus hardfork, the
+Operator Fee Feature was [proposed][proposed] and approved. The feature
+was considered to be a part of the upcoming Isthmus hardfork. Rust
+implementations of OP Stack components went ahead and implemented the
+feature. For months, the feature sat idle. After it was eventually merged,
+there were issues in the execution client changes for the feature. Removing
+implementations of the feature was considered, but the effort to remove it
+was not seen as worthwhile. After a while, the feature was finally
+complete. It effectively alone delayed the hardfork by a month or so.
+
+This could have been avoided by being able to implement the feature using
+a feature-specific toggle in the `RollupConfig`. The feature toggle would
+be toggled on in the [`superchain-registry`][scr] once the feature is ready.
+In this way, features could be implemented and left in client software
+without being connected to hardfork. The key part: *no changes would
+need to be made to client software to enable the feature*, only enabling
+the feature inside the [`superchain-registry`].
+
+
+# Problem Statement
+
+Bundling features within hardforks limits flexibility:
+
+- All features in a hardfork activate at once, preventing selective enablement
+  for testing or debugging.
+- Custom chains or testnets cannot easily remap or delay individual features
+  without forking entire client implementations.
+- CLI overrides are limited to hardfork timestamps, not per-feature adjustments.
+- The set of features for a given hardfork must be determined prior to
+  implementation, locking and committing to scope.
+
+This rigidity hinders iteration, increases deployment risks, and complicates
+superchain governance.
+
+
+# Current Approach
+
+- Hardforks are defined by named timestamps in the `RollupConfig`
+  (e.g., `fjordTime: uint64`).
+- Node logic checks `timestamp >= hardforkTime` to enable the all features
+  in the hardfork.
+- Features are implicitly tied to their hardfork; no per-feature granularity.
+- CLI overrides exist for hardfork timestamps (e.g., `--override.fjord=123456`
+  in `op-node`).
+    - Overrides in the `op-node` exist [here][op-overrides]. In the `kona-node`,
+      [here][kona-overrides].
+
+
+# Proposed Solution
+
+Decouple features from hardforks by modifying `RollupConfig`s to include a
+mapping from feature names to hardfork names (e.g., a `map[Feature]Fork`
+where the key is feature like "operatorFee" and value is a hardfork like "isthmus").
+If a feature is absent from the map, it remains inactive.
```

Comment:
Is this done during development and then the set of features in a hard fork and hard coded when they are shipped (either in the release that sets the activation timestamp for testnet or mainnet or post-main net activation)? Or is the intent to keep the individual feature toggles forever and continue to have to list them in the rollup config?